### PR TITLE
KT-22861 "Add annotation target" quick fix is not available on annotation with use site target

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-22861
https://youtrack.jetbrains.com/issue/KT-22862

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddAnnotationTargetFix.kt [hunk 1]
```diff
@@ -81,14 +84,30 @@ private fun KtAnnotationEntry.getRequiredAnnotationTargets(annotationClass: KtCl
     }.flatten().toSet()
 
     val annotationTargetValueNames = AnnotationTarget.values().map { it.name }
-    return (requiredTargets + otherReferenceRequiredTargets).filter { it.name in annotationTargetValueNames }
+    return (requiredTargets + otherReferenceRequiredTargets).toSet().filter { it.name in annotationTargetValueNames }
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/AddAnnotationTargetFix.kt]: Probably `distinct()` is better here

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/AnnotationChecker.kt [hunk 2]
```diff
@@ -221,9 +221,18 @@ class AnnotationChecker(
             }.toSet()
         }
 
-        fun getDeclarationSiteActualTargetList(annotated: KtElement, descriptor: ClassDescriptor?, trace: BindingTrace):
+        fun getDeclarationSiteActualTargetList(
```
Comment (mglukhikh) [inline compiler/frontend/src/org/jetbrains/kotlin/resolve/AnnotationChecker.kt]: I don't like an idea to change front-end code in this commit: as result we get an additional parameter used only from IDE code. I think it's better to call `getActualTargetList` from fix (make it public for this), and then take necessary targets from this list (effectively inlining your function here). As alternative idea, you can write separate function in IDE for a case with `all = true`.

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddAnnotationTargetFix.kt [hunk 3]
```diff
@@ -81,14 +84,30 @@ private fun KtAnnotationEntry.getRequiredAnnotationTargets(annotationClass: KtCl
...
+
+    if (useSiteTarget == null) {
+        return AnnotationChecker.getDeclarationSiteActualTargetList(annotatedElement, null, BindingTraceContext())
+
+    } else {
+        val target = KotlinTarget.USE_SITE_MAPPING[useSiteTarget?.getAnnotationUseSiteTarget()] ?: return emptyList()
+
+        val actualTargetList =
+            AnnotationChecker.getDeclarationSiteActualTargetList(annotatedElement, null, BindingTraceContext(), all = true)
+        if (actualTargetList.none { it == target }) return emptyList()
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/AddAnnotationTargetFix.kt]: `if (target !in actualTargetList)`

**Comment by t-kameyama** [comment]
Fixed

**Comment by mglukhikh** [comment]
I think now it's Ok, I will recheck and merge it. Thank you.

**Comment by mglukhikh** [comment]
Polished a bit and merged, thank you.

**Pull request closed without merging**